### PR TITLE
Aurora and Theresa Scrabble

### PR DESCRIPTION
**Pull request by auroralemieux**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   How do you feel you and your partner did in sharing responsibilities?	|   We both improved our understanding of the separation of driver/navigator responsibilities. Traded off in 20 minute increments. That worked pretty well.	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  Mainly because the scoring was based on data (individual letter scores) that won't ever change. So it's something that you can access for all scrabble game instances and the functionality is the same.  |
|   Describe an example of a test you wrote for a _nominal case_. |    To test that the self.score method was working correctly, we manually added up the score for a particular word and used must_equal to verify that passing in that word to .score outputted the correct score.	|
|   Describe an example of a test you wrote for an _edge case_. |   For the Dictionary class, we checked to make sure the build_list method grabbed what it should -- that it only contained words less than 8 letters, and that it pulled the first and the last words that it should have from our txt file.	|
|   What was your final test coverage percentage?	|   100%. Bam. (Which really means nothing.)	|
|   Is there a specific piece of code you'd like feedback on?	|   Are we writing the right tests? (not are they working)	|

### CODE REVIEW

File: specs/scoring_specs.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
+require_relative 'spec_helper'
+require_relative '../lib/scoring.rb'
+
+describe "Scoring is a class" do
+  it "must be type of scoring" do
+    scoring = Scrabble::Scoring.new
```
Comment (CheezItMan) [inline specs/scoring_specs.rb:6]: Since we're not making any Scoring Objects (it only has class methods) we don't need to test any Scoring instances (Scoring.new).

File: specs/scoring_specs.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    scoring.must_be_instance_of Scrabble::Scoring
+  end
+
+  describe "initialize" do
```
Comment (CheezItMan) [inline specs/scoring_specs.rb:10]: Don't have an `initialize` method so we don't need this describe.

File: specs/scoring_specs.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    it "Raise ArgumentError if array contains anything but string" do
+      proc {Scrabble::Scoring.highest_score_from(["cat", 1, "blah"])}.must_raise ArgumentError
+    end
+
+    it "Return should be a String" do
+      Scrabble::Scoring.highest_score_from(%w(cat dog piglet oink)).must_be_kind_of String
+    end
+
+    it "picks the 7-letter word in a tie" do
+      Scrabble::Scoring.highest_score_from(%w(qqqqqj daaaaaa i)).must_equal "daaaaaa"
```
Comment (CheezItMan) [inline specs/scoring_specs.rb:72]: In each of these test cases you should also vary the order to verify it's picking the correct word in a tie.

File: specs/tilebag_spec.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+require_relative '../lib/tilebag.rb'
+
+describe "Tilebag is a class" do
+
+  before do
+    @tilebag = Scrabble::TileBag.new
+  end
+
+  it "must be type of TileBag" do
+    tilebag = Scrabble::TileBag.new
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb:11]: Why this if you have the `before` block

File: specs/tilebag_spec.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      end
+    end
+
+    it "Returns an array of string with length num" do
+      returned_tiles_array = @tilebag.draw_tiles(2)
+      returned_tiles_array.length.must_equal 2
+    end
+
+    it "Updates @tile_set correctly" do
+      original = Scrabble::TileBag.new
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb:40]: If you wanted to save the original size you should probably simply do:

`original_length = @tilebag.tile_set.length`

And then at the end 

`@tilebag.tile_set.length.must_equal original_length - 3`

No need to make another instance of TileBag

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

|  Feature 	|   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions 	|   Well done, I'll check through your tests and answer on it at the bottom.  	|
|   Both Teammates contributed to the codebase	|   It looks like you cooperated pretty well, trading off on a schedule.	|
|   Regular Commits with meaningful commit messages 	|   Pretty good commit messages and plenty of them	|
|  **Scoring class**	|   	|
|  `score` method	|   Nicely done!	|
|   Uses appropriate data structure to store the letter score	|   Good use of a hash	|
|   Appropriately handles edge cases	|   Good checks for invalid data	|
|   Tests `score` edge cases (one-letter word, 7 letter word)	|   Handles 7-letter word scores	|
|  `highest_scoring_word` method	|   Nice use of Enumerables	|
|   Appropriately handles edge cases	|   Checks for highest scoring 7-letter word, and otherwise goes for the smallest, highest scoring word, should also test the same list of words in different orders.	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|   Checks, ties, no words and no array	|
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|   Check	|
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|   Check	|
|  Returns `true` or `false` from `won?` method |  Check, nice use of a ternary operator |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |   Check, nicely done with the validation method  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|   Good use of the Array, and I like the method to generate the list of tiles.	|
|  `draw_tiles` method uses the parameter to draw a set of random tiles |   Check	|
| `tiles_remaining` method returns a count	|   Check	|
|  Tests `draw_tiles` edge cases for different parameter inputs	|   Nicely done	|


## Summary

Nicely done.  I think you did a pretty good job of selecting things to test.  One thing to recommend is if you're selecting an element or elements from an Array, ensuring that order doesn't affect the answer.  I would also for testing Strings, test empty strings (you did), invalid strings (you did), one-letter strings (nope), and in this case 7-letter words and words beyond 7-letters.

**Review by CheezItMan — COMMENTED** [review]
I've left some comments in your PR code.
